src/models/document/base-document-content.ts

```diff
@@ -0,0 +1,1081 @@
+                self.mergeRow(row, rowInfo);
+              }
+            }
+            else {
+              // entire row is being moved to a new row
+              if ((rowInsertIndex < rowIndex) || (rowInsertIndex > rowIndex + 1)) {
+                self.moveRowToIndex(rowIndex, rowInsertIndex);
+              }
+            }
+          }
+          else {
+            if ((rowDropLocation === "left") || (rowDropLocation === "right")) {
+              // part of row is being moved to an existing row
+              self.moveTilesToExistingRowAtIndex(rowTiles, rowInfo);
+            }
+            else {
+              // part of row is being moved to a new row
+              self.moveTilesToNewRowAtIndex(rowTiles, rowInsertIndex);
+            }
+          }
+        });
+      }
+    }
+  }))
+  .actions(self => ({
+    addSharedModel(sharedModel: SharedModelType) {
+      // we make sure there isn't an entry already otherwise adding a shared
+      // model twice would clobber the existing entry.
+      let sharedModelEntry = self.sharedModelMap.get(sharedModel.id);
+
+      if (!sharedModelEntry) {
+        sharedModelEntry = SharedModelEntry.create({sharedModel});
+        self.sharedModelMap.set(sharedModel.id, sharedModelEntry);
+      }
+
+      return sharedModelEntry;
+    },
+    addSharedModelFromImport(id: string, sharedModelEntry: SharedModelEntrySnapshotType){
+      if (self.sharedModelMap){
+        self.sharedModelMap.set(id, sharedModelEntry);
+      }
+    }
+  }))
+  .actions(self => ({
+    userAddTile(toolId: string, options?: IDocumentContentAddTileOptions) {
+      const result = self.addTile(toolId, options);
+      const newTile = result?.tileId && self.getTile(result.tileId);
+      if (newTile) {
+        logTileDocumentEvent(LogEventName.CREATE_TILE, { tile: newTile });
+      }
+      return result;
+    },
+    userDeleteTile(tileId: string) {
+      const tile = self.getTile(tileId);
+      if (tile) {
+        logTileDocumentEvent(LogEventName.DELETE_TILE, { tile });
+        self.deleteTile(tileId);
+      }
+    },
+    userMoveTiles(tiles: IDragTileItem[], rowInfo: IDropRowInfo) {
+      tiles.forEach(tileItem => {
+        const tile = self.getTile(tileItem.tileId);
+        tile && logTileDocumentEvent(LogEventName.MOVE_TILE, { tile });
+      });
+      self.moveTiles(tiles, rowInfo);
+    },
+    userCopyTiles(tiles: IDropTileItem[], rowInfo: IDropRowInfo) {
+      const dropRow = (rowInfo.rowDropIndex != null) ? self.getRowByIndex(rowInfo.rowDropIndex) : undefined;
+      const results = dropRow?.acceptTileDrop(rowInfo)
+                      ? self.copyTilesIntoExistingRow(tiles, rowInfo)
+                      : self.copyTilesIntoNewRows(tiles, rowInfo.rowInsertIndex);
+      self.logCopyTileResults(tiles, results);
+      return results;
+    },
+    handleDragCopyTiles(dragTiles: IDragTilesData, rowInfo: IDropRowInfo) {
```

Comment:
Should this function be moved to `drag-tiles.ts` or renamed?

Reply:
Yeah I think several functions should be moved from BaseDocumentContentModel. I just didn't want to cause more conflicts, feel free to move functions as you get shared model copying working in master.